File: dart/exercises-(in_kotlin)/exercise3_drawable_shapes.kt
// ============================================================
// SE 3242: Android Application Development
// Week 2 - Kotlin Exercises
// Exercise 3: Drawable Shapes with Interfaces
// Student: Precious [Last Name]
// Date: March 2026
// Description: Demonstrates interfaces, multiple interface
//              implementation, and polymorphism in Kotlin
//              using geometric shapes.
// ============================================================

import kotlin.math.PI

/**
 * Represents any object that can be drawn to the console.
 *
 * Classes that implement [Drawable] must supply a visual representation
 * via [draw] and identify themselves via [shapeType].
 */
interface Drawable {

    /**
     * Renders an ASCII-art representation of the shape to standard output.
     */
    fun draw()

    /**
     * Returns the name of this shape (e.g. "Circle", "Square", "Triangle").
     *
     * @return A human-readable string identifying the shape type.
     */
    fun shapeType(): String
}

/**
 * Represents any object whose geometric measurements can be computed.
 *
 * Classes that implement [Measurable] must provide an [area] and a [perimeter].
 */
interface Measurable {

    /**
     * Calculates and returns the area of the shape.
     *
     * @return The area expressed in square units.
     */
    fun area(): Double

    /**
     * Calculates and returns the perimeter (or circumference) of the shape.
     *
     * @return The perimeter expressed in linear units.
     */
    fun perimeter(): Double
}

// ── Concrete shape classes ────────────────────────────────────────────────────

/**
 * A geometric circle implementing both [Drawable] and [Measurable].
 *
 * @property radius The radius of the circle in abstract units.
 */
class Circle(val radius: Double) : Drawable, Measurable {

    /** Prints a simple ASCII circle to the console. */
    override fun draw() {
        println("  Shape: Circle (radius = $radius)")
        println("     *  *  *  ")
        println("  *           *")
        println(" *             *")
        println("  *           *")
        println("     *  *  *  ")
    }

    /**
     * Returns the area of the circle: $A = \pi r^2$.
     *
     * @return Area in square units.
     */
    override fun area(): Double = PI * radius * radius

    /**
     * Returns the circumference of the circle: $C = 2\pi r$.
     *
     * @return Circumference in linear units.
     */
    override fun perimeter(): Double = 2 * PI * radius

    /** @return The string `"Circle"`. */
    override fun shapeType(): String = "Circle"
}

/**
 * A geometric square implementing both [Drawable] and [Measurable].
 *
 * @property side The length of one side in abstract units.
 */
class Square(val side: Double) : Drawable, Measurable {

    /** Prints a simple ASCII square to the console. */
    override fun draw() {
        println("  Shape: Square (side = $side)")
        println("  +-------+")
        println("  |       |")
        println("  |       |")
        println("  |       |")
        println("  +-------+")
    }

    /**
     * Returns the area of the square: $A = s^2$.
     *
     * @return Area in square units.
     */
    override fun area(): Double = side * side

    /**
     * Returns the perimeter of the square: $P = 4s$.
     *
     * @return Perimeter in linear units.
     */
    override fun perimeter(): Double = 4 * side

    /** @return The string `"Square"`. */
    override fun shapeType(): String = "Square"
}

/**
 * A geometric triangle implementing both [Drawable] and [Measurable].
 *
 * The area formula uses the classic base-height formula, while the perimeter
 * assumes an equilateral-style model where all sides equal [side].
 *
 * @property base   The length of the base in abstract units.
 * @property height The perpendicular height from the base in abstract units.
 * @property side   The length of one side used for the perimeter calculation.
 */
class Triangle(val base: Double, val height: Double, val side: Double) : Drawable, Measurable {

    /** Prints a simple ASCII triangle to the console. */
    override fun draw() {
        println("  Shape: Triangle (base = $base, height = $height, side = $side)")
        println("       *       ")
        println("      * *      ")
        println("     *   *     ")
        println("    *     *    ")
        println("   *       *   ")
        println("  ***********  ")
    }

    /**
     * Returns the area of the triangle: $A = \frac{1}{2} \times base \times height$.
     *
     * @return Area in square units.
     */
    override fun area(): Double = 0.5 * base * height

    /**
     * Returns the perimeter of the triangle: $P = 3 \times side$.
     *
     * @return Perimeter in linear units.
     */
    override fun perimeter(): Double = 3 * side

    /** @return The string `"Triangle"`. */
    override fun shapeType(): String = "Triangle"
}

// ── Higher-order utility ──────────────────────────────────────────────────────

/**
 * Applies a given [action] to every [Drawable] in [shapes].
 *
 * This higher-order function abstracts iteration so callers can supply any
 * lambda without knowing the internal list mechanics.
 *
 * @param shapes A list of [Drawable] objects to operate on.
 * @param action A lambda that receives each [Drawable] and performs some work.
 */
fun applyToShapes(shapes: List<Drawable>, action: (Drawable) -> Unit) {
    shapes.forEach { shape -> action(shape) }
}

// ── Entry point ───────────────────────────────────────────────────────────────

/**
 * Entry point for Exercise 3.
 *
 * Demonstrates:
 * - Creating a `List<Drawable>` with [Circle], [Square], and [Triangle]
 * - Iterating with `forEach` to call [draw] on each shape
 * - Casting to `List<Measurable>` and printing [area] and [perimeter]
 * - Using `filterIsInstance<Circle>()` to extract only circles
 * - Calling [applyToShapes] with a lambda that prints each shape's type
 */
fun main() {
    // ── Part 1: Draw all shapes ───────────────────────────────────────────────
    println("=== Part 1: Drawing all shapes ===\n")

    val shapes: List<Drawable> = listOf(
        Circle(radius = 5.0),
        Square(side = 4.0),
        Triangle(base = 6.0, height = 4.0, side = 5.0)
    )

    shapes.forEach { shape ->
        shape.draw()
        println()
    }

    // ── Separator ─────────────────────────────────────────────────────────────
    println("------------------------------------------------------------")
    println()

    // ── Part 2: Print area and perimeter for each shape ───────────────────────
    println("=== Part 2: Measurements (area & perimeter) ===\n")

    @Suppress("UNCHECKED_CAST")
    val measurables = shapes as List<Measurable>   // safe: all shapes implement Measurable

    measurables.forEach { m ->
        // Re-cast to Drawable just to get the name for the label
        val name = (m as Drawable).shapeType()
        println("  $name")
        println("    Area      : ${"%.4f".format(m.area())} sq units")
        println("    Perimeter : ${"%.4f".format(m.perimeter())} units")
        println()
    }

    // ── Separator ─────────────────────────────────────────────────────────────
    println("------------------------------------------------------------")
    println()

    // ── Part 3: Filter only circles ───────────────────────────────────────────
    println("=== Part 3: Circles only (filterIsInstance) ===\n")

    val circles = shapes.filterIsInstance<Circle>()

    if (circles.isEmpty()) {
        println("  No circles found.")
    } else {
        println("  Found ${circles.size} circle(s):")
        circles.forEach { c ->
            println("    → Circle with radius ${c.radius}, area = ${"%.4f".format(c.area())}")
        }
    }

    println()
    println("------------------------------------------------------------")
    println()

    // ── Part 4: applyToShapes with a lambda ───────────────────────────────────
    println("=== Part 4: Shape types via applyToShapes() ===\n")

    applyToShapes(shapes) { shape ->
        println("  Shape type: ${shape.shapeType()}")
    }

    println()
    println("Exercise 3 complete.")
}

File: dart/exercises-(in_dart)/exercise3_drawable_shapes.dart
// ============================================================
// SE 3242: Android Application Development
// Week 2 - Dart Exercises
// Exercise 3: Drawable Shapes with Interfaces
// Student: [Your Name]
// Date: March 2026
// Description: Demonstrates interfaces (abstract classes),
//              multiple interface implementation, and
//              polymorphism in Dart using geometric shapes.
// ============================================================

import 'dart:math';

// ─────────────────────────────────────────────────────────────
// Interfaces (abstract classes)
// ─────────────────────────────────────────────────────────────

/// Interface for any object that can be drawn to the console.
///
/// Implementors must provide [draw] and [shapeType].
abstract class Drawable {
  /// Prints an ASCII representation of the shape to the console.
  void draw();

  /// Returns the name of the shape type (e.g. `"Circle"`).
  String shapeType();
}

/// Interface for any geometric object whose area and perimeter
/// can be calculated.
///
/// Implementors must provide [area] and [perimeter].
abstract class Measurable {
  /// Returns the area of the shape.
  double area();

  /// Returns the perimeter (circumference) of the shape.
  double perimeter();
}

// ─────────────────────────────────────────────────────────────
// Concrete shape classes
// ─────────────────────────────────────────────────────────────

/// A circle defined by its [radius].
///
/// Implements both [Drawable] and [Measurable].
class Circle implements Drawable, Measurable {
  /// The radius of the circle.
  final double radius;

  /// Creates a [Circle] with the given [radius].
  Circle(this.radius);

  /// Prints an ASCII representation of a circle.
  @override
  void draw() {
    print('  Circle (r = $radius)');
    print('     .oOo.    ');
    print('   oO     Oo  ');
    print('  O         O ');
    print('   oO     Oo  ');
    print('     .oOo.    ');
  }

  /// Returns the area: π × radius².
  @override
  double area() => pi * radius * radius;

  /// Returns the circumference: 2 × π × radius.
  @override
  double perimeter() => 2 * pi * radius;

  /// Returns `"Circle"`.
  @override
  String shapeType() => 'Circle';
}

/// A square defined by its [side] length.
///
/// Implements both [Drawable] and [Measurable].
class Square implements Drawable, Measurable {
  /// The length of one side of the square.
  final double side;

  /// Creates a [Square] with the given [side] length.
  Square(this.side);

  /// Prints an ASCII representation of a square.
  @override
  void draw() {
    print('  Square (side = $side)');
    print('  +--------+');
    print('  |        |');
    print('  |        |');
    print('  +--------+');
  }

  /// Returns the area: side².
  @override
  double area() => side * side;

  /// Returns the perimeter: 4 × side.
  @override
  double perimeter() => 4 * side;

  /// Returns `"Square"`.
  @override
  String shapeType() => 'Square';
}

/// An equilateral triangle defined by its [base], [height], and [side].
///
/// For an equilateral triangle all three sides are equal, so [side]
/// is used for the perimeter calculation. [base] and [height] are
/// used for the area calculation.
///
/// Implements both [Drawable] and [Measurable].
class Triangle implements Drawable, Measurable {
  /// The length of the base of the triangle.
  final double base;

  /// The perpendicular height of the triangle.
  final double height;

  /// The length of one side (equilateral — all sides equal).
  final double side;

  /// Creates a [Triangle] with the given [base], [height], and [side].
  Triangle({required this.base, required this.height, required this.side});

  /// Prints an ASCII representation of a triangle.
  @override
  void draw() {
    print('  Triangle (base = $base, height = $height, side = $side)');
    print('      /\\      ');
    print('     /  \\     ');
    print('    /    \\    ');
    print('   /______\\   ');
  }

  /// Returns the area: 0.5 × base × height.
  @override
  double area() => 0.5 * base * height;

  /// Returns the perimeter: 3 × side (equilateral).
  @override
  double perimeter() => 3 * side;

  /// Returns `"Triangle"`.
  @override
  String shapeType() => 'Triangle';
}

// ─────────────────────────────────────────────────────────────
// Higher-order helper
// ─────────────────────────────────────────────────────────────

/// Applies [action] to every shape in [shapes].
///
/// This is a custom higher-order function — [action] is a lambda
/// (first-class function value) supplied by the caller, giving
/// full control over what happens to each shape.
///
/// Example — print each shape type:
/// ```dart
/// applyToShapes(shapes, (s) => print(s.shapeType()));
/// ```
void applyToShapes(List<Drawable> shapes, void Function(Drawable) action) {
  shapes.forEach(action);
}

// ─────────────────────────────────────────────────────────────
// Entry point
// ─────────────────────────────────────────────────────────────

/// Demonstrates interfaces, multiple-interface implementation,
/// and polymorphism through a series of shape operations.
void main() {
  final separator = '─' * 50;

  // ── Create a mixed list of shapes ─────────────────────────
  final List<Drawable> shapes = [
    Circle(5.0),
    Square(4.0),
    Triangle(base: 6.0, height: 5.196, side: 6.0),
  ];

  // ── Section 1: forEach → draw() each shape ────────────────
  print(separator);
  print('Section 1 — Draw each shape (forEach + draw())');
  print(separator);
  shapes.forEach((shape) {
    shape.draw();
    print('');
  });

  // ── Section 2: cast to Measurable → area & perimeter ──────
  print(separator);
  print('Section 2 — Area & perimeter (Measurable interface)');
  print(separator);

  // Every shape in our list also implements Measurable, so the
  // cast is safe here.  In production code you would use
  // whereType<Measurable>() instead.
  final measurables = shapes.cast<Measurable>();
  measurables.forEach((m) {
    // Recover the shape name via the Drawable interface.
    final name = (m as Drawable).shapeType();
    print('$name → area: ${m.area().toStringAsFixed(4)}'
        '  |  perimeter: ${m.perimeter().toStringAsFixed(4)}');
  });

  print('');

  // ── Section 3: whereType<Circle>() ────────────────────────
  print(separator);
  print('Section 3 — Circles only (whereType<Circle>)');
  print(separator);
  shapes
      .whereType<Circle>()
      .forEach((c) => print('Circle with radius ${c.radius} '
          '→ area: ${c.area().toStringAsFixed(4)}'));

  print('');

  // ── Section 4: higher-order applyToShapes ─────────────────
  print(separator);
  print('Section 4 — Shape types via applyToShapes (higher-order)');
  print(separator);
  applyToShapes(shapes, (shape) => print('Shape type: ${shape.shapeType()}'));

  print(separator);
}
